tests/CMakeLists.txt

```diff
@@ -110,6 +110,19 @@ add_executable(sycl_tests
 # Also test instant submission mode
 target_compile_definitions(sycl_tests PRIVATE -DACPP_ALLOW_INSTANT_SUBMISSION=1)
 target_include_directories(sycl_tests PRIVATE ${Boost_INCLUDE_DIRS} ${CMAKE_CURRENT_SOURCE_DIR} ${OpenMP_CXX_INCLUDE_DIRS})
+
+if (ACPP_CUDA_BACKEND_AVAILABLE AND (NOT ACPP_TARGETS OR ACPP_TARGETS MATCHES "cuda" OR ACPP_TARGETS MATCHES "generic"))
+  message(STATUS "CUDA enabled - searching for CUDA toolkit")
+  find_package(CUDAToolkit QUIET)
+  if (CUDAToolkit_FOUND)
+    target_link_libraries(sycl_tests PRIVATE CUDA::cudart)
```

Comment:
I'm not really convinced this is the right approach. I understand that you want to test whether interop with custom operation works with CUDA.
But the whole point of `generic` is to decouple the generated binary from backends in order to have backend-independent binaries. The code should be able to run (minus interop, of course) independently of whether the binary is directly linked against CUDA or not.

This PR would change our CI to no longer test this if a CUDA toolkit is found. In this case, CUDA would be linked directly to the application, and we would be unable to detect symbol creep from the CUDA backend into the application.

So, for SSCP I think we should at most make this an optional test, if we want to have it at all. Testing both cases would require building the tests twice - once with CUDA linked and once without, and I'm not sure this double stress on CI is worth it.
Also, I'm not sure why we should treat CUDA differently here than say, HIP (where we have even less CI resources).

Maybe we should have some dedicated interop test suite which could then link against backend libraries as much as it wants. This test suite could be fairly small, and hence only require little resources and be run on all backends that support interop.

This change might be out of scope for this PR, so I would be happy to maintain the status quo of interop testing with SSCP for now until we do it properly in a future PR. Would this be okay for you?

Reply:
Totally on board. Honestly, the large number of macros and defines and targets and their interaction still confuses me. 

> I understand that you want to test whether interop with custom operation works with CUDA

Indeed, well, I was surprised to learn there was already a test for this, but it wasn't running. I suspect there's something I'm just not understanding here, but..

As best I can tell, before this PR the tests would never run, since the macro `ACPP_LIBKERNEL_COMPILER_SUPPORTS_CUDA` is basically the same as `__CUDACC__`. 

The [CUDA docs say](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/)

> Defined when compiling CUDA source files

The [Clang docs say](https://llvm.org/docs/CompileCudaWithLLVM.html)

> Both clang and nvcc define __CUDACC__ during CUDA compilation

Both of which to my mind are ambiguous, but I assume mean compiling `.cu` files regardless of host or device pass.  So the `custom_enqueue` tests would have only run if treating `.cpp` files as `.cu` files?  Perhaps this is the case for the `cuda` targets of `acpp`, but not for the `generic` target, hence my confusion?


> Also, I'm not sure why we should treat CUDA differently here than say, HIP (where we have even less CI resources).

Yeah I was reluctant to touch HIP since I have no way to develop/test this locally.


> I would be happy to maintain the status quo of interop testing with SSCP for now until we do it properly in a future PR. Would this be okay for you?

This is fine.  So to be clear, I can drop all commits apart from the last one?  This will mean my change is not tested (yet), at least with the `generic` target.

Reply:
> Totally on board. Honestly, the large number of macros and defines and targets and their interaction still confuses me.

Yeah, flexibility inevitably pulls in complexity. There's no other SYCL compiler that supports as many compilation flows as AdaptiveCpp.

> Indeed, well, I was surprised to learn there was already a test for this, but it wasn't running. I suspect there's something I'm just not understanding here, but..
> As best I can tell, before this PR the tests would never run, since the macro ACPP_LIBKERNEL_COMPILER_SUPPORTS_CUDA is basically the same as __CUDACC__.

Yes. `ACPP_LIBKERNEL_COMPILER_SUPPORTS_CUDA` basically tests whether we are in `--acpp-targets=cuda`, *not* `--acpp-targets=generic`.
This might have been deliberate to avoid the the questions around linking of `libcudart` that you discovered :)

> Perhaps this is the case for the cuda targets of acpp, but not for the generic target, hence my confusion?

Yes, exactly. `--acpp-targets=cuda` runs through the clang CUDA toolchain, so in this case you are effectively compiling your SYCL code with a CUDA compiler in disguise that was augmented by AdaptiveCpp to also understand SYCL. The same is true for `--acpp-targets=hip` and the clang HIP toolchain respectively.

On the other hand, with `--acpp-targets=generic` everything is controlled by AdaptiveCpp and there are no other programming models involved.

The `cuda`/`hip` targets integrate tightly with backend models CUDA/HIP and in particular also support mix-and-match of CUDA/HIP code with SYCL. As such, they are primarily useful when migrating code.

The advantage of `generic` is that, since it's completely under the control of AdaptiveCpp, it offers much more features and AdaptiveCpp-specific performance optimizations. Since this compiler is a JIT compiler with a backend-unified code representation, binary portability is much better. After all, every binary can dispatch to all backends depending on what hardware it finds in the system at runtime.

> Yeah I was reluctant to touch HIP since I have no way to develop/test this locally.

Understandable. We do have AMD hardware in CI, but of course iterating on code while relying solely on CI for testing is cumbersome.

> This is fine. So to be clear, I can drop all commits apart from the last one? This will mean my change is not tested (yet), at least with the generic target.

Yes, this would be my suggestion for now.

Reply:
That all makes total sense, thanks for the detailed explanation.

I've dropped the other commits, rebased on `develop`, and re-worded the remaining commit message to elaborate on the questions raised here.